[MM-25673] Stabilize master

ethervoid opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

Multiple changes to stabilize gorp:

- Fix tests
- Partially revert #3 to avoid breaking the `Rename` and some `Select` function cases
- Include CircleCI configuration
- Make Ginkgo work

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

https://mattermost.atlassian.net/browse/MM-25673

ethervoid committed:
Fix ginkgo problem

ethervoid committed:
Fix tests

ethervoid committed:
mod tidy

ethervoid committed:
wip

ethervoid committed:
Fix more

ethervoid committed:
test wip

ethervoid committed:
Fix rename and select error

ethervoid committed:
Fix NullTime

ethervoid committed:
Fix unique index tests

ethervoid committed:
Fix test waiting for non-fatal error on select

ethervoid committed:
Fix testing script

ethervoid committed:
Add circleci configuration

ethervoid commented:
/update-branch

ethervoid committed:
Remove travis

ethervoid committed:
More circleci

ethervoid committed:
Fix test script

ethervoid committed:
Revert "Fix test waiting for non-fatal error on select"

This reverts commit 8458910326afd4c8d707efe87f4ac370b054b21d.

ethervoid committed:
Skip that test because we removed that functionallity

ethervoid committed:
Bump to go 1.14.4

ethervoid committed:
Fixed Rename functionallity

Now we're able to use the Rename functionallity and also use hyphen
columns as required by our code in Mattermost

lieut-data commented on gorp.go:
```diff
@@ -257,15 +257,15 @@ func columnToFieldIndex(m *DbMap, t reflect.Type, name string, cols []string) ([
 			cArguments := strings.Split(field.Tag.Get("db"), ",")
 			fieldName = cArguments[0]
 
+			if fieldName == "" || fieldName == "-" {
```
With respect to the partial revert, can you elaborate on the impact to the bot fields in question?

ethervoid replied:
Should be none and we recover the Rename feature

crspeller replied:
This should be fine. We were not using the rename feature at the time so I didn't worry about breaking it. It keeps the change acutally needed from the other PR.

lieut-data reviewed (commented):
One comment, but shifting review to @crspeller given the impact to his original changes.

agnivade commented on gorp_test.go:
```diff
@@ -48,9 +48,10 @@ var (
 )
 
-func init() {
+func TestMain(m *testing.M) {
```
:+1:

agnivade commented on go.mod:
```diff
@@ -3,11 +3,11 @@ module github.com/mattermost/gorp
 go 1.13
 
 require (
-	github.com/go-gorp/gorp v2.2.0+incompatible
```
:+1:

agnivade commented on gorp.go:
```diff
@@ -257,15 +257,15 @@ func columnToFieldIndex(m *DbMap, t reflect.Type, name string, cols []string) ([
 			cArguments := strings.Split(field.Tag.Get("db"), ",")
 			fieldName = cArguments[0]
 
+			if fieldName == "" || fieldName == "-" {
```
Just wanted to clarify quickly: is this change already covered by the change in the unit tests? If not, can we add some tests to cover them?

ethervoid replied:
Yes it is covered but I’ll add more explicit tests to be really sure

agnivade reviewed (commented):
Really nice work! Just some small comments.

agnivade commented:
Requesting a CI review from Elisabeth.

metanerd commented on .circleci/config.yml:
```diff
@@ -0,0 +1,28 @@
+          name: Run Tests
+          command: |
+            go get -u github.com/onsi/ginkgo/ginkgo
+            ./test_all.sh
```
In this bash script there are also remnants of travis and our blacklisting does not recognize the name `test_all.sh`. Personally I would favour converting it into a Makefile to go with the standard, but we can also do this at a later iteration.

ethervoid replied:
Ok, I’ll tackle it

ethervoid replied:
Done

metanerd reviewed (changes requested):
Thanks for taking care!

ethervoid committed:
Tests for hyphen columns

We check that hypehn columns are returned, even in their default
form, but is not possible to insert or update data on them

ethervoid committed:
Makefile and test parameters

- Created Makefile and removed the test script
- Deprecate the user of env. variables in favor or flags to test
  different databases

ethervoid committed:
Fix test for sqlite

agnivade reviewed (approved):
Fantastic

metanerd reviewed (approved):
Great! :+1:
